### ISSUE TITLE
Introducing a let-expression causes type error

### ISSUE THREAD

**Issue by mschlaipfer** [dafny-lang/dafny#314]
The following example works fine.

```dafny
datatype S = S(G: array<int>)
datatype T = T(F: array<S>, ghost Repr: set<object>)

predicate Valid(t: T)
  reads t.F
  reads (set i | 0 <= i < t.F.Length :: t.F[i].G)
{
  t.Repr == (set i | 0 <= i < t.F.Length :: t.F[i].G) + {t.F}
}
```

When I introduce a let-expression `var short := t.F` and replace `t.F` as in

```dafny
datatype S = S(G: array<int>)
datatype T = T(F: array<S>, ghost Repr: set<object>)

predicate Valid(t: T)
  reads t.F
  reads (set i | 0 <= i < t.F.Length :: t.F[i].G)
{
  var short := t.F;
  t.Repr == (set i | 0 <= i < short.Length :: short[i].G) + {short}
}
```

I get a type error:
```
> $ Dafny.exe lang-test47.dfy
Dafny 2.3.0.10506
lang-test47.dfy(13,44): Error: type of right argument to + (set<array?<S>>) must agree with the result type (set<object>) (covariant type parameter would require array?<S> <: object)
1 resolution/type errors detected in lang-test47.dfy
```

**Comment by mschlaipfer** [comment]
The same happens when t.F is renamed via pattern matching such as in
```
match t {
    case T(F, Repr) => Repr == (set i | 0 <= i < F.Length :: F[i].G) + {F}
}
```
fails

**Comment by davidcok** [comment]
cf Issue #321

**Comment by davidcok** [comment]
Similar fix to #701. Fixed in PR #734.
